plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3Service.java

```diff
@@ -238,104 +265,141 @@ protected PasswordAuthentication getPasswordAuthentication() {
                             }
                         });
                     }
-                    clientConfiguration.getApacheHttpClientConfig()
-                        .setSslSocketFactory(createSocksSslConnectionSocketFactory(clientSettings.proxySettings.getAddress()));
+                    clientBuilder.socketFactory(createSocksSslConnectionSocketFactory(clientSettings.proxySettings.getAddress()));
                 });
             } else {
-                if (clientSettings.proxySettings.getType() != ProxySettings.ProxyType.DIRECT) {
-                    clientConfiguration.setProxyProtocol(clientSettings.proxySettings.getType().toProtocol());
-                }
-                clientConfiguration.setProxyHost(clientSettings.proxySettings.getHostName());
-                clientConfiguration.setProxyPort(clientSettings.proxySettings.getPort());
-                clientConfiguration.setProxyUsername(clientSettings.proxySettings.getUsername());
-                clientConfiguration.setProxyPassword(clientSettings.proxySettings.getPassword());
+                clientBuilder.proxyConfiguration(buildHttpProxyConfiguration(clientSettings));
             }
         }
 
-        if (Strings.hasLength(clientSettings.signerOverride)) {
-            clientConfiguration.setSignerOverride(clientSettings.signerOverride);
+        clientBuilder.socketTimeout(Duration.ofMillis(clientSettings.readTimeoutMillis));
+
+        return clientBuilder;
+    }
+
+    static ProxyConfiguration buildHttpProxyConfiguration(S3ClientSettings clientSettings) {
+        ProxyConfiguration.Builder proxyConfiguration = ProxyConfiguration.builder();
+        if (clientSettings.proxySettings.getType() == ProxySettings.ProxyType.SOCKS) {
```

Comment:
Hm ... this method is never called for `ProxySettings.ProxyType.SOCKS` because of https://github.com/opensearch-project/OpenSearch/pull/7372/files?diff=split&w=1#diff-b8fbfee9d71f7f2caf16514727692e968b4638afcafce12b4698e2046e6cb673R255, do we really need it?

Reply:
You are right, we don't need it for production scenarios, but we have some tests in place which verify the `ProxyConfiguration` being built here, so we can either take a call to remove the [SOCKS proxy configuration test](https://github.com/opensearch-project/OpenSearch/pull/7372/files?diff=split&w=1#diff-b37c837e0049aa779e99e9eb61047212f346432f9d16b256b86182b5d0890097R312), which would make sense since this method is never called for that scenario, but the reason I kept it is that it asserts that a proxy configuration is not built for SOCKS proxy. Otherwise we need to have this check.

Reply:
May be go with `Optional` instead, since we don't set any `ProxyConfiguration` for SOCKS:

```
static Optional<ProxyConfiguration> buildHttpProxyConfiguration(S3ClientSettings clientSettings) {
 if (clientSettings.proxySettings.getType() == ProxySettings.ProxyType.SOCKS) {
    return Optional.empty();
 }

 ...
}
```